Code is both Python 2.7 and 3.5 compliant

olaurino opened the pull request:
# Description

Resolve #76.

Introduce support for Python 3 while keeping code base compatible with Python 2.7.

Introduced a Python 3.5 travis build, but we can target more Python 3.x versions later.

This PR also tries and introduce new unit and integration tests.

As a side effect, this PR also fixes: #186 #241

DougBurke commented:
@olaurino are you looking for comments at this stage, or is more work needed before a review is needed?

olaurino commented:
Comments are always welcome, especially in this case where the PR is going to be massive. You already caught an unintended change, so that's good :)

DougBurke commented on sherpa/astro/datastack/ds.py:
```diff
@@ -28,6 +27,10 @@
+try:
+    import stk
+except:
+    logger.warning("could not import stk library. CIAO stack files and syntax will be disabled")
```
Is the idea here that, because the only use of the `stk` module is already within a try/except block, there is no need to create a "replacement" routine for `stk.build()` in the except clause? I thought that the change looked odd, until I checked into how the `stk` module was used here. Do we need to document this (e.g. in a comment) for anyone who wants to use `stk,build` in this module? Or am I being over sensitive?

olaurino replied:
That is exactly what I had in mind. Yes, we could improve the documentation (as a matter of fact I had a comment that explained that but apparently I dropped it after changing the `try..catch` block. If the stack library is not present and you use the stack syntax (e.g. `@filename`) you'll get a file not found error, which paired with the warning should be reasonable enough.

DougBurke replied:
I noted that there looked to be similar-ish logic in the `sherpa/astro/datastack/utils.py` module. I didn't look closely enough to see if the code was close enough to be able to remove one of them. That is, set `Datastack.load_pha` to `load_wrapper(ui.load_pha)`.

Perhaps there's something obviously different that I'm just missing (as I'm only skimming through the code).

olaurino replied:
You are absolutely right, I noticed that too. I have the temptation to fix things like this as I go along, but I don't want that to be in the way of the python 3 migration, and the clock is ticking. There does not seem to be any reason for the duplication, but there is some complexity in the way the wrapping is done, so that might hide a trap... or I just forgot to remove a method when I did the refactoring for #22. Either way, I would rather spend the time required to figure this out adding test coverage where we are missing it. In any case I appreciate the level of your review, so please keep making these kinds of comments :)

DougBurke commented:
I see that the python 2.7 tests are now back to passing (yay), but the stk and group tests are skipped. Is the idea to leave these for now and concentrate on the main code (which looks like replacing lots of use of izip!) and then come back to these later (in part because these are OTS from our view, so we are waiting for other teams to update them)?

olaurino commented:
@DougBurke that's the idea.

bsipocz commented on .travis.yml:
```diff
@@ -19,6 +19,8 @@ addons:
 # We use matrix include so that xspec builds (which are more complicated and long) are started first
 matrix:
+  allow_failures:
+      - python: "3.5"
```
This should be exactly the same as the build you want to allow to fail, so here I guess:
```
     - env: XSPECVER="12.9.0i" FITS="astropy" INSTALL_TYPE=develop TEST=submodule MATPLOTLIBVER=1.5
       python: "3.5"
       sudo: required
       dist: trusty
```

olaurino replied:
Thanks @bsipocz. Not sure why I kept the `allow_failures` line, since I am happy with the build to be red until all the tests pass. I'll remove that line altogether, rather than fixing it.

DougBurke commented on helpers/develop.py:
```diff
@@ -1,3 +1,4 @@
+from __future__ import print_function
 #
 #  Copyright (C) 2015  Smithsonian Astrophysical Observatory
```
Copyright year needs updating ;-)

DougBurke commented on sherpa/__init__.py:
```diff
@@ -80,12 +81,12 @@ def get_config():
     # If NOSHERPARC is set, read in system config file
     # ignore any user config file
-    if (os.environ.has_key('NOSHERPARC') == True):
+    if (('NOSHERPARC' in os.environ) == True):
```
This (and the following case) can be further cleaned up to

    if 'NOSHERPARC' in os.environ:

DougBurke commented on sherpa/astro/flux.py:
```diff
@@ -97,8 +98,8 @@ def simulated_pars_within_ranges(mysamples, mysoftmins, mysoftmaxs):
-        print '%s = %g, + %g, - %g' % (title, arg[0], arg[1] - arg[0],
-                                       arg[0] - arg[2])
+        print('%s = %g, + %g, - %g' % (title, arg[0], arg[1] - arg[0],
+                                       arg[0] - arg[2]))
```
Is it worth changing

    fmt % vals

to

    fmt.format(vals)

[with appropriate changes like %<xx> -> {<yy>} in `fmt`]? I had thought (a long while ago) that the % syntax was deprecated but I'm a lot-less sure now.

DougBurke commented on sherpa/optmethods/optfcts.py:
```diff
@@ -89,58 +91,56 @@ def _narrow_limits( myrange, xxx, debug ):
     myxmin = raise_min_limit( myrange, xmin, x, debug=False )
     myxmax = lower_max_limit( myrange, x, xmax, debug=False )
 
     if False != debug:
```
I guess you don't want to do cleanup like change this to

    if debug:

I guess this might not be quite the same, since debug need not be a boolean, but earlier checks in this function rely on `if debug` so it should be safe here.

DougBurke commented on sherpa/optmethods/optfcts.py:
```diff
@@ -537,9 +537,9 @@ def stat_cb0( pars ):
+        seed = random.randint(0, 2147483648) # pow(2,31) == 2147483648L
     if population_size is None:
-        population_size = max( population_size, 12 * x.size )
+        population_size = 12 * x.size
```
Why was the `max()` check removed here? Oh. hold on, population_size is by definition None here, so I see why you did it.

Comment left for anyone else who didn't spot this (ie me when I look at this again).

DougBurke commented on sherpa/utils/__init__.py:
```diff
@@ -24,16 +25,16 @@
 import operator
 import inspect
-from itertools import izip
+from six.moves import zip as izip
```
Can this be moved to be with the other imports from `six.moves`? It's a minor stylistic point, so only worth doing if you feel it is worthwhile.

DougBurke commented on sherpa/utils/__init__.py:
```diff
@@ -369,10 +380,6 @@ def requires_pylab(test_function):
 eps = numpy.finfo(numpy.float32).eps
 
 
-def erfinv(y):
```
Was this removed because it ends up not being used anywhere in sherpa? It might be a useful function to have for users (but would need documentation and tests and a commitment from us it's worth keeping).

DougBurke commented on sherpa/utils/tests/test_utils_unit.py:
```diff
@@ -0,0 +1,203 @@
+#
+#  copyright (c) 2016  smithsonian astrophysical observatory
```
Another "picky" comment: what has happened to all the capital letters in this section?

DougBurke commented:
I think, once we've got the python 3.5 travis builds green we should add in tests for python versions 3.3. and 3.4. Ideally we'd be able to support all versions, but if not, it would be good to know which ones we don't support.

DougBurke commented:
@olaurino - I've played around with a few fixes but you may not want them (I am just relying on the existing tests rather than checking if new tests are needed). You can find them at

https://github.com/DougBurke/sherpa/tree/python3

and they are basically StringIO, basestring, and xrange fixes.

DougBurke commented:
@olaurino - OOPS. I did not mean to push those changes! I'm not sure what the best way to fix this is, so I'm going to leave it for now.

olaurino commented:
@DougBurke you can leave them here, thanks! I wonder where the conflicts are coming from, I'll look into that and rebase.

olaurino commented:
Rebasing, original tip: 7842371

DougBurke commented:
Did I do something wrong to cause the conflict? I have more changes over on https://github.com/DougBurke/sherpa/tree/python3 which I could push (after sorting out any conflict), but that can't happen until this evening at the earliest.

I expect the python 2.7 + xspec tests to fail on a6f0534 - if so, there's a fix in the commits on my tree.

olaurino commented:
> I expect the python 2.7 + xspec tests to fail on a6f0534 - if so, there's a fix in the commits on my tree.

Yes, I would prefer to fix this and get the python2 tests to pass.

My only concern with the rest of the commits on your branch is that they do not seem exactly in the scope of the python3 migration, is that true?

> Did I do something wrong to cause the conflict?

I don't think so, the conflict was "soft": the kind of conflicts git does not seem to care about (I didn't have to do anything during the rebase) but github doesn't like.

DougBurke commented:
> My only concern with the rest of the commits on your branch is that they do not seem exactly in the scope of the python3 migration, is that true?

The changes I made are all for python3 conversion I believe (well, there maybe a few PEP8 changes along the way, and given that I did accidentally merge to the wrong account some care may be needed). Can you be more explicit as to which changes you mean?

olaurino commented:
For some reason I must have landed on a different branch on your fork. I can see that the changes are in the scope of the python3 work, sorry for the confusion. Please feel free to push them to this branch. Thanks!

olaurino commented:
I rebased the commits from @DougBurke's `python3` branch and I am pushing them here.

DougBurke commented:
Thanks for the rebase @olaurino ; I don't have time for Sherpa at work today ;-)

DougBurke commented:
I started looking at the ds9 code, and have a simple commit at https://github.com/sherpa/sherpa/tree/python3-ds9 which now at least runs the test when ds9 is installed. The tests all fail, but better than claiming that ds9 is not available.

```
% python setup.py test -a sherpa/image/tests/test_image.py 
...
        try:
            imager.showArray(arr)
        except:
>           raise DS9Err('noimage')
E           sherpa.utils.err.DS9Err: Could not display image

sherpa/image/ds9_backend.py:86: DS9Err
=========================== 7 failed in 8.24 seconds ===========================
```

I wonder if we should augment the DS9Err message to include the exception message to make debugging these easier. A more useful error message is

```
In [1]: import numpy as np

In [2]: from sherpa.image import DS9

In [3]: img = np.arange(300*400).reshape(300, 400)

In [4]: imager = DS9.DS9Win("foo", False)

In [5]: imager.doOpen()

In [6]: imager.showArray(img)
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-6-3b8253b00644> in <module>()
----> 1 imager.showArray(img)

/home/djburke/sherpa/sherpa3/sherpa/image/DS9.py in showArray(self, arr, **kargs)
    551         self.xpaset(
    552             cmd='array [%s]' % (_formatOptions(arryDict),),
--> 553             dataFunc=arr.tofile,
    554         )
    555 

/home/djburke/sherpa/sherpa3/sherpa/image/DS9.py in xpaset(self, cmd, data, dataFunc)
    646             dataFunc=dataFunc,
    647             template=self.template,
--> 648             doRaise=self.doRaise,
    649         )
    650 

/home/djburke/sherpa/sherpa3/sherpa/image/DS9.py in xpaset(cmd, data, dataFunc, template, doRaise)
    335     try:
    336         if dataFunc:
--> 337             dataFunc(p.stdin)
    338         elif data:
    339             p.stdin.write(data)

OSError: obtaining file position failed
```

and it looks like it is using bytes for communication, or at least that's what I take from seeing the b before the string below:

```
In [7]: imager.xpaget("frame all")
Out[7]: b'1 \n'
```

DougBurke commented on sherpa/models/model.py:
```diff
@@ -46,12 +46,12 @@ def cache_model(cls, pars, xlo, *args, **kwargs):
         if use_caching:
 
-            data = [ numpy.array(pars).tostring(), str(kwargs.get('integrate',0)), 
+            data = [ numpy.array(pars).tostring(), kwargs.get('integrate', b'0'),
```
So, if `integrate` is defined in `kwargs`, it is likely to be True or False. Is that not going to cause a problem in line 54 below - e.g. in python 2.7 I get

```
>>> a = ["foo", True, "bar"]
>>> b''.join(a)
TypeError: sequence item 1: expected string, bool found
```

Or have we guaranteed that by the time we call this routine we are guaranteed that the integrate keyword is a string?

olaurino replied:
Hmm... you are probably right. I'll try and see if there is a test passing `integrate` as a boolean. Probably there is not. Or I am missing something :)

olaurino replied:
the keyword is actually set in several tests (implicitly by the model classes themselves), but the potential bug would only be triggered by models that enable caching *and* have `integrate=True`, which is never the case in the tests. I am not even sure what the expected behavior in should be in this corner case.

Also, this looks like a potential bug in the current code as well, right @DougBurke?

DougBurke replied:
I think the current code is okay, since it has the str() around the get call, but it isn't ideal: it really should be mapping all inputs to a "sensible" serialization scheme (e.g. 0 or 1). Or at least, that's how I read it. I know I have not really tried to use the caching code, so it wouldn't surprise me if there's some odd corner cases and not enough testing.

olaurino replied:
````   
bmap = {True: b'1', False: b'0'}
data = [numpy.array(pars).tostring(), bmap.get(kwargs.get('integrate', False), b'0'),
                    numpy.asarray(xlo).tostring()]
````

This should work in all cases: it always returns `b'0' or b'1'`, including the case in which `integrate` is not a boolean.

DougBurke replied:
Fancy! Perhaps I'm overthinking it with regard to how the integrate value is serialized, but I'm happy for the above to go in, with some testing  ;-)

DougBurke commented:
@olaurino I have a commit to push to fix test_datastack - you don't see it here because it needs the stk module available, but I've enabled this module in my python 2.7 test suite. I didn't want to push it to this branch in case it messed up your work. It's available at https://github.com/DougBurke/sherpa/commit/0f199d65a6cdcd4b005a5f1cb72bbbb9cacaf6c0 or I can push it if you give the go-ahead.

DougBurke commented:
I had a quick look at the XSPEC tests and they are all being skipped, even when XSPEC is being built. I ended up with this error message - is it expected at this time?

```
In [1]: from sherpa.astro import xspec
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-13a81ec87068> in <module>()
----> 1 from sherpa.astro import xspec

/home/djburke/sherpa/sherpa3/sherpa/astro/xspec/__init__.py in <module>()
     23 from sherpa.models import Parameter, ArithmeticModel, modelCacher1d
     24 from sherpa.models.parameter import hugeval
---> 25 import sherpa.astro.xspec._xspec
     26 from sherpa.utils import guess_amplitude, param_apply_limits
     27 from sherpa.astro.utils import get_xspec_position

ImportError: dynamic module does not define module export function (PyInit__xspec)
```

olaurino commented:
I noticed that the xspec tests were being skipped earlier today. I left the office while the tests were running. I'll push the commit that fixes it tomorrow.

DougBurke commented:
The xspec tests almost pass now: yay.

The two failures are

a) whether the model name is reported as starting with 'xs' or not - e.g. https://travis-ci.org/sherpa/sherpa/jobs/147759623#L5056 - and is an issue I've seen before when working on the XSPEC tests. There is obviously something funky going on somewhere in the internals (e.g. I note that it looks like the python 2.7 tests pass here).

b) The type error for a numeric conversion - e.g. https://travis-ci.org/sherpa/sherpa/jobs/147759623#L5163 - looks like it'll be fun (this is new to me so I assume an actual Python 2<->3 conversion issue).

DougBurke commented:
yay - you've got python 3.5 passing!

DougBurke commented:
I've cherry-picked the datastack fix that was on my fork of sherpa and pushed it to the https://github.com/sherpa/sherpa/tree/python3-datastack branch, so it doesn't get lost (and I can check that travis thinks it is okay).

DougBurke commented on sherpa/astro/xspec/__init__.py:
```diff
@@ -196,7 +201,7 @@ def __init__(self, filename, name='xstbl', parnames=(),
 
         # make translation table to turn reserved characters into '_'
         bad = string.punctuation+string.whitespace
```
Should this line not be changed too? 

It's not clear to me why this commit the problem (e.g. `xstablemodel.tmod` != `tablemodel.tmod` at https://travis-ci.org/sherpa/sherpa/jobs/147759623#L5056) since I would have expected to see AttributeError's instead. Or are the errors being caught and ignored somewhere?

olaurino replied:
`punctuation` and `whitespace` are still attributes of `sting` in Python3, as far as I can tell. The problem with `(xs)tablemodel` is that the code tries to instantiate the xspec table model class, and falls back to the regular table model if it doesn't work, silently. The Python 3 migration is a lot of work already, and I am trying to be strict and resist the urge to fix the issues I stumble upon.

DougBurke replied:
Aha - I had forgotten to do an `import string` when I tested. I'd also forgotten about the I-think-should-be-deprecated code saying

```
    try:
        xspec-table-model-code()
    except:
        sherpa-table-model-code()
```

olaurino commented on sherpa/image/DS9.py:
```diff
@@ -333,12 +333,12 @@ def xpaset(cmd, data=None, dataFunc=None, template=_DefTemplate,
-        if dataFunc:
-            dataFunc(p.stdin)
-        elif data:
+        if isinstance(data, six.string_types):
```
In order to fix the Python 3 incompatibilities, which are related to the use of `numpy.ndarray.tofile`, I had to create the data with `tobytes` on line 553: f3b3a11b4553779cdc379a16f29bf5e4R553.

In principle, we might do without `dataFunc`, since users can pass the data already as a string or byte array, as done here. However, if there are any users using this argument (I seriously doubt that) chances are their code is broken in Python 3 just as ours was.

However, I'll wait for a round of review: we could simply deprecate the `dataFunc` argument so we don't have to change the API. We probably need to slightly update the doc string, though.

DougBurke commented on sherpa/image/DS9.py:
```diff
@@ -333,8 +333,16 @@ def xpaset(cmd, data=None, dataFunc=None, template=_DefTemplate,
+        try:  # Python 2 bytes (which are actually strings)
+            unicode(data, "ascii")  # unicode does not exist in Python 3
             data = bytearray(data, "UTF-8")
+        except:
```
Should we only catch the error here rather than catching anything (in case `unicode` in 2.7 raises anything), that is, use

    except NameError:

Similarly, it would be better to just wrap the `unicode` line to minimize problem cases, but perhaps that is too excessive?

olaurino replied:
I agree with catching `NameError`, that's what I usually do (but not in this case, apparently). However, this is just a less ugly (and maybe less explicit, so not even better) way of expressing a
````
if six.PY3:

else:
````

block.

DougBurke commented on sherpa/astro/io/crates_backend.py:
```diff
@@ -934,13 +934,13 @@ def get_pha_data(arg, make_copy=True, use_background=False):
         pha = phadataset.get_crate(phadataset.get_current_crate())
-        if (pha.get_key('HDUCLAS1').value == 'SPECTRUM' or
-                pha.get_key('HDUCLAS2').value == 'SPECTRUM'):
+        if (pha.get_key('HDUCLAS1').value == b'SPECTRUM' or
```
Eek. Is crates really returning `bytes` rather than `str` here? That doesn't seem sensible (the domain is a better map to the FITS standard than unicode, but it's not perfect, and I think it's likely to confuse people). I'll have to see what astropy.io.fits does.

Would it be worth changing these `<crate>.get_key(...).value` calls to use `_try_key` and get it to convert from bytes to str, so that if crates changes behavior we (may) have less places to change?

olaurino replied:
I am trying to get the code to work, but yes, in some cases I agree that crates should be returning strings and bury this ugliness down a couple of layers :)

And yes, the refactoring you suggest makes sense either way.

DougBurke commented on sherpa/astro/datastack/tests/test_datastack.py:
```diff
@@ -680,7 +680,7 @@ def test_case_7(self):
         assert f == [1, 2]
 
-        f = datastack.query_by_obsid('7867')
+        f = datastack.query_by_obsid(7867)
```
The current doc string for `query_by_obsid` startes that the value can be a str or an integer. Does the str option no-longer work (and if so, how much work would it be to restore)?

olaurino replied:
I just realized we didn't have a test for the integer case, so I changed one of the tests. A string is tested before and after this test. Or am I missing something?

DougBurke replied:
Ahhh, looking at the test as a whole I see there's two calls to this method, and we now test both cases. Phew and sorry for the false alarm.

DougBurke replied:
I should point out that I'm jet-lagged, so about the only thing I feel up to is nit-picking [1] in a code review ;-)

[1] not sure of the equivalent Italian expression

DougBurke commented on extern/Makefile.in:
```diff
@@ -1,8 +1,9 @@
-# Makefile.in generated by automake 1.14.1 from Makefile.am.
+# Makefile.in generated by automake 1.11.1 from Makefile.am.
```
How come we are now using an older version of automake? I guess as this is external code we (as in sherpa) doesn't care as long as it works, but 1.11 seems rather old.

olaurino replied:
Because I didn't bother to run `autoreconf` in a virtual machine with `ubuntu` 14 or 15. It does not matter too much, but we should probably rerun `autoreconf` from a decently new distro when dust have settled.

DougBurke commented on extern/region-4.9/configure:
```diff
@@ -1,6 +1,6 @@
 #! /bin/sh
 # Guess values for system-dependent variables and create Makefiles.
-# Generated by GNU Autoconf 2.63 for region 4.8.
+# Generated by GNU Autoconf 2.63 for region 4.7.
```
Why is this saying "region 4.7" (and, more importantly, the PACKAGE_VERSION/STRING lines below also say 4.7)?

olaurino replied:
Because that's the code I got from Mark. Most likely I changed that manually last time I incorporated changes from the `region` library, but I didn't do it this time.

There are a number of changes I have to make to these scripts because despite the efforts they still contain a number of dependencies on CIAO Makefiles, or mistakes like this. In principle, I should just be given a tar ball created with `make distcheck`, but in practice that's not the case.

DougBurke replied:
It would be nice to have these fixed by upstream, and not you ;-)

DougBurke commented on sherpa/astro/data.py:
```diff
@@ -839,7 +839,7 @@ def ignore_bad(self):
     def _dynamic_group(self, group_func, *args, **kwargs):
 
-        keys = kwargs.keys()[:]
+        keys = list(kwargs.keys())[:]
```
I believe `list` returns a copy of the iterable, so we can drop the trailing `[:]` from such cases. See https://docs.python.org/3/library/stdtypes.html#list
(this is in a number of places throughout the code base).

DougBurke commented:
I have a small commit to fix the `inspect.getargspec` deprecation warnings from `sherpa/sim/mh.py` (which appear even with the recent change to `sherpa/conftest.py`, presumably because the tests in `sherpa/astro/sim/tests/test_astro_sim.py` currently fail). It's on my fork and is at

https://github.com/DougBurke/sherpa/tree/python3-sim-getargspec

Although not noted in the commit message, the code could be made a bit cleaner by using `sherpa.utils.get_keyword_defaults`, but at present the sim module doesn't require any parts of Sherpa, so I decided to keep this "feature" (I know that there is some updated code for pyblocxs somewhere, but we haven't discussed whether it makes sense to update the sherpa version).

DougBurke commented:
We must not have any tests of `sherpa.astro.ui.sample_flux`, since the following fails (there is also the fact that once it passes there'll be some numpy Deprecation Warnings):

# use of `xrange`

```
% git rev-parse HEAD
63d86d5d0bdcde19de6d4d3a7c538f503601a9cb

% ipython --pylab
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
Using matplotlib backend: Qt4Agg

In [1]: from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find xpaget on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: ui.load_pha('sherpa-test-data/sherpatest/3c273.pi')
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/3c273.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file sherpa-test-data/sherpatest/3c273.arf
read RMF file sherpa-test-data/sherpatest/3c273.rmf
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/3c273_bg.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file sherpa-test-data/sherpatest/3c273_bg.pi

In [3]: ui.set_source(ui.powlaw1d.pl)

In [4]: ui.fit()
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
Dataset               = 1
Method                = levmar
Statistic             = chi2gehrels
Initial fit statistic = 8.15838e+10
Final fit statistic   = 50.5351 at function evaluation 25
Data points           = 46
Degrees of freedom    = 44
Probability [Q-value] = 0.231118
Reduced statistic     = 1.14852
Change in statistic   = 8.15838e+10
   pl.gamma       1.72329     
   pl.ampl        0.000147339 

In [5]: ui.plot_fit()

In [6]: ui.sample_flux()
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-6-165be028d0d9> in <module>()
----> 1 ui.sample_flux()

<string> in sample_flux(modelcomponent, lo, hi, id, num, scales, correlated, numcores, bkg_id, Xrays, confidence)

/lagado.real/sherpa/sherpa3/sherpa/astro/ui/utils.py in sample_flux(self, modelcomponent, lo, hi, id, num, scales, correlated, numcores, bkg_id, Xrays, confidence)
  11519                                               correlated=correlated,
  11520                                               numcores=numcores,
> 11521                                               bkg_id=bkg_id)
  11522 
  11523         return sherpa.astro.flux.calc_sample_flux(id, lo, hi, self, fit, data,

/lagado.real/sherpa/sherpa3/sherpa/astro/ui/utils.py in sample_energy_flux(self, lo, hi, id, num, scales, correlated, numcores, bkg_id)
  11381                                              sherpa.astro.utils.calc_energy_flux,
  11382                                              correlated, num, lo, hi, numcores,
> 11383                                              scales)
  11384 
  11385     # DOC-NOTE: are scales the variance or standard deviation?

/lagado.real/sherpa/sherpa3/sherpa/astro/flux.py in sample_flux(fit, data, src, method, correlated, num, lo, hi, numcores, samples)
     78     hardmins = fit.model._get_thawed_par_hardmins()
     79     hardmaxs = fit.model._get_thawed_par_hardmaxes()
---> 80     samples = within_limits(samples, hardmins, hardmaxs)
     81 
     82     return calc_flux(fit, data, src, samples, method, lo, hi, numcores)

/lagado.real/sherpa/sherpa3/sherpa/astro/flux.py in within_limits(mysamples, mymins, mymaxs)
     57         num_par = mysamples.shape[1]
     58         for row in mysamples:
---> 59             for index in xrange(num_par):
     60                 if row[index] < mymins[index]:
     61                     row[index] = mymins[index]

NameError: name 'xrange' is not defined

```
# deprecation warnings

If you try the above on the master branch with python 2.7 you get the following (only showing the output of `ui.sample_flux` since the previous commands are the same):

```
In [6]: f = ui.sample_flux()
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
sherpa/astro/flux.py:148: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  upconfidence = sf[(1.0 - myconfidence) * size - 1]
sherpa/astro/flux.py:149: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  loconfidence = sf[myconfidence * size - 1]
original model flux = 1.19924e-12, + -1.19468e-13, - 1.19468e-13
model component flux = 1.19924e-12, + -1.19468e-13, - 1.19468e-13

```

DougBurke commented:
I've got an update to `setup.py` to include support for 3.5 - it's in the `python3-setup` branch - i.e. https://github.com/sherpa/sherpa/tree/python3-setup

olaurino commented:
@DougBurke could you open PRs with your changes against this branch? That would help me tracking and merging them. I am sure I am forgetting a branch you set up already! :)

olaurino commented:
BTW @DougBurke 	33c606d should fix the issue you reported.

In the latest volley of commits I added some tests, compensated for issues they uncovered, and refactored some workarounds introduced to deal with crates data so a) they are easier to change back if crates returns strings rather than bytes, and b) all crates and astropy tests should now pass.

DougBurke commented:
I can confirm that with the latest master I can now run sample_flux in python 3.5. Yay.

DougBurke commented:
I have adder PRs #242 and #243 for the changes I know haven't been integrated into the Python3 branch. I think the other branches I have lying around have been integrated (or you've made the changes yourself); I'll have to check them.

DougBurke commented:
Just to note that I've looked through the various other branches (ie other than the two used for PRs 242 and 243) I have lying around and I believe they are all included already in this branch (and so have removed them from the repository).

DougBurke commented:
I've added PR #245 which is an annoying "add 2016" to the copyright years sweep of this commit.

DougBurke commented:
I've noticed we have a bunch of tests of the form `self.assertTrue(a is b)` (many of them have been converted in this PR from using `self.assert_`. It would be nice to have these as `self.assertIs(a, b)`, to get better diagnostics and documentation of the tests, but is it worth doing it in this PR?

DougBurke commented:
From discussions on #245, it looks like we should remove `sherpa/include/capsulethunk.h` since we don't aim to support python 2.6 or 3.0/1, so can just deal with Capsule objects. Or is that too much work at this time?

DougBurke commented:
I've just added PR #250 which removes the capsulethunk file. It was surprisingly easy (yay for encapsulation, I guess), so maybe I missed something?

DougBurke commented:
Should we note that this has been deprecated by #252 (or close this in favor of 252)?

olaurino commented:
I am about to merge this and rebase #252!

cdeil commented:
Thank you!!!

I'll add a Python 3 build for Gammapy in the next days.

Will there be a Python 3-compatible Sherpa conda package for testing soon, or only after the next stable Sherpa release?

olaurino commented:
@cdeil the conda package will most likely come in a few weeks with the 4.8.2 release. However, the conda recipes are [included in this repo](https://github.com/sherpa/sherpa/tree/master/recipes/conda), so you could produce them yourself. They may require bumping versions and other adjustments to the dependencies every now and then (in this case you certainly need to update the python version), but if you can build Sherpa from sources, it should be a straightforward process. If that's not true, or you have issues or questions, please let me know.

I have quite a packed schedule, but if I find some time I'll publish some test cut of the python 3 conda package. I just can't promise, so please don't hold your breath :wink:

olaurino commented:
By the way, I created a `sherpa 4.8py3` conda package (with the `dev` label) about a month ago to support https://github.com/astropy/saba testing. It may be out of date, and it's Linux only:
http://anaconda.org/sherpa/sherpa/files

taldcroft commented:
Congratulations on this tremendous accomplishment!!  :tada:  This removes the last real blocker from Chandra operations tools migrating to Python 3.
